crypto/mem.c

```diff
@@ -250,6 +250,7 @@ void CRYPTO_free(void *str, const char *file, int line)
     }
 
     free(str);
+    str = NULL;
```

Comment:
What are your intentions here? Your change as it is does not make much sense: The pointer `str` is a local variable and its value is not used afterwards anymore. So a decent compiler will optimize this assignment away anyway.

Reply:
This doesn't do anything as @mspncp notes.

Adding something like this to the macro wrapper `OPENSSL_free` might work.

Reply:
Such a change  of the `OPENSSL_free()` function(-like macro) would would be a breaking change of its semantics: Effectively, you are now passing the pointer itself by reference, nly lvalues can now be passed as arguments to `OPENSSL_free()` and the function argument itself is modified. No C-programmer would expect such a behaviour. (C++ programmers wouldn't be so much surprised, but the semantics for the `OPENSSL_free()` call still would be weird.) Of course, @xnox could write  his own `MYOPENSSL_free()` macro if he wants this behaviour.

Reply:
I never said it was a good idea :)

Reply:
@mspncp yeah, changing the function signature to `OPENSSL_free(void **str)` indeed will break all users.

Basically what I dislike is that a couple of times we keep on having use-after-free and double-free bugs that would be prevented if the APIs would not only decrement use count, and free things when they reach zero, but allow nullify the pointer. I.e. especially with engines, which have been freed/deallocated multiple times recently, leading to segfaults of not just the engine, but the libssl library / the whole process.

So yeah, no idea how to fix it up, whilst having things C-like, and not break ABI.